File: packages/processor/go/reader/description_inferrer.go
// @docspec:module {
//   id: "docspec-go-reader-inferrer",
//   name: "Description Inferrer",
//   description: "Generates human-readable descriptions from Go identifier names using verb prefix decomposition and camelCase splitting, enabling Tier 0 zero-config documentation.",
//   since: "3.0.0"
// }
package reader

import (
	"strings"
	"unicode"
)

var verbMap = map[string]string{
	"Get": "Retrieves", "Find": "Finds", "Fetch": "Fetches", "Load": "Loads",
	"Create": "Creates", "New": "Creates a new", "Add": "Adds", "Insert": "Inserts",
	"Update": "Updates", "Modify": "Modifies", "Set": "Sets",
	"Delete": "Deletes", "Remove": "Removes",
	"Validate": "Validates", "Check": "Checks", "Verify": "Verifies",
	"Process": "Processes", "Handle": "Handles", "Execute": "Executes",
	"Send": "Sends", "Publish": "Publishes", "Emit": "Emits",
	"Convert": "Converts", "Transform": "Transforms", "Parse": "Parses",
	"Calculate": "Calculates", "Compute": "Computes", "Count": "Counts",
	"Is": "Checks whether", "Has": "Checks if has", "Can": "Determines if can",
}

// InferDescription generates a description from a Go identifier name.
//
// @docspec:method { since: "3.0.0" }
// @docspec:deterministic
// @docspec:intentional "Infers a natural-language description from a camelCase function/type name using a verb-to-sentence mapping"
func InferDescription(name string) string {
	words := splitCamelCase(name)
	if len(words) == 0 {
		return ""
	}

	verb := words[0]
	rest := strings.Join(words[1:], " ")

	if prefix, ok := verbMap[verb]; ok {
		if rest != "" {
			return prefix + " " + strings.ToLower(rest) + "."
		}
		return prefix + " the resource."
	}

	humanized := strings.Join(words, " ")
	return strings.ToUpper(humanized[:1]) + humanized[1:] + "."
}

// @docspec:deterministic
func splitCamelCase(s string) []string {
	var words []string
	var current strings.Builder
	for i, r := range s {
		if i > 0 && unicode.IsUpper(r) {
			if current.Len() > 0 {
				words = append(words, current.String())
				current.Reset()
			}
		}
		current.WriteRune(r)
	}
	if current.Len() > 0 {
		words = append(words, current.String())
	}
	return words
}

File: packages/processor/rust/src/reader/description_inferrer.rs
//! Infer descriptions from Rust identifiers.
//!
//! @docspec:module {
//!   id: "docspec-rust-reader-inferrer",
//!   name: "Description Inferrer",
//!   description: "Generates human-readable description strings from snake_case Rust identifiers by matching verb prefixes against a known dictionary and constructing natural-language sentences.",
//!   since: "3.0.0"
//! }

use std::collections::HashMap;
use std::sync::LazyLock;

static VERB_MAP: LazyLock<HashMap<&str, &str>> = LazyLock::new(|| {
    let mut m = HashMap::new();
    m.insert("get", "Retrieves"); m.insert("find", "Finds"); m.insert("fetch", "Fetches");
    m.insert("create", "Creates"); m.insert("new", "Creates a new");
    m.insert("add", "Adds"); m.insert("insert", "Inserts"); m.insert("save", "Saves");
    m.insert("update", "Updates"); m.insert("modify", "Modifies"); m.insert("set", "Sets");
    m.insert("delete", "Deletes"); m.insert("remove", "Removes");
    m.insert("validate", "Validates"); m.insert("check", "Checks"); m.insert("verify", "Verifies");
    m.insert("process", "Processes"); m.insert("handle", "Handles"); m.insert("execute", "Executes");
    m.insert("send", "Sends"); m.insert("publish", "Publishes"); m.insert("emit", "Emits");
    m.insert("convert", "Converts"); m.insert("transform", "Transforms"); m.insert("parse", "Parses");
    m.insert("calculate", "Calculates"); m.insert("compute", "Computes"); m.insert("count", "Counts");
    m.insert("is", "Checks whether"); m.insert("has", "Checks if has");
    m
});

/// @docspec:deterministic
/// @docspec:intentional "Transforms a snake_case identifier into a human-readable sentence using verb-prefix dictionary lookup"
pub fn infer(name: &str) -> String {
    let words: Vec<&str> = name.split('_').collect();
    if words.is_empty() { return String::new(); }

    let verb = words[0];
    let rest: Vec<&str> = words[1..].to_vec();
    let rest_str = rest.join(" ");

    if let Some(prefix) = VERB_MAP.get(verb) {
        if !rest_str.is_empty() {
            format!("{} {}.", prefix, rest_str)
        } else {
            format!("{} the resource.", prefix)
        }
    } else {
        let humanized = words.join(" ");
        let mut chars = humanized.chars();
        match chars.next() {
            None => String::new(),
            Some(c) => format!("{}{}.", c.to_uppercase(), chars.collect::<String>()),
        }
    }
}
